_posts/2022-09-28-import-big-csv.md

```diff
@@ -0,0 +1,306 @@
+  "~/datasets/my-data"
+  )
+```
+
+It takes about the same amount of time as scanning the CSV files (0.02
+seconds).
+
+Counting the number of occurences for each of the 10 unique values in
+a column takes just 1 second. One second to summarize 140 million
+rows. It is a little slower than performing this query when the entire
+dataset is loaded in memory but loading the files is much faster.
+
+One of the advantages of the Arrow ecosystem is that the approach that
+works with R also works with Python. And because both language use the
+same C++ backend, the code looks very similar.
+
+
+## Single file API in Python
+
+There are two functions in the Python Single API to read CSV files:
+`read_csv()` and `open_csv()`. While `read_csv()` loads all the data
+in memory, and does it fast by using multiple threads to read
+different part of the files, `open_csv` reads the data in batches and
+uses a single thread.
+
+If the CSV file is small enough, you should use `read_csv`. The code
+to read the CSV file and write it to a Parquet file would then look
+like this:
+
+```python
+import pyarrow as pa
+import pyarrow.csv
+import pyarrow.parquet as pq
+
+in_path = '~/datasets/data.csv'
+out_path = '~/datasets/data.parquet'
+
+data =  pa.csv.read_csv(in_path)
+
+pq.write_table(data, out_path)
+```
+
+In our case, the file is too large to fit in memory. So instead of
+using `read_csv()`, we need to use `open_csv()`. Because, the CSV file
+is read in chunks, the code is a little more complex. We loop through
+each chunk, read it, and write it to the Parquet file. This uses
+little memory but can be a little slow as a single thread is used to
+read the file. When using `open_csv()`, the data types need to be
+consistent in your columns. The inferrence is made on the first chunk,
+and if the type changes halfway through your dataset, you will run
+into errors. You can avoid this by specifying the data types manually.
+
+```python
+# from https://stackoverflow.com/a/68563617/1113276
+import pyarrow as pa
+import pyarrow.parquet as pq
+import pyarrow.csv
+
+in_path = '~/datasets/data.csv'
+out_path = '~/datasets/data.parquet'
+
+writer = None
+with pyarrow.csv.open_csv(in_path) as reader:
+    for next_chunk in reader:
+        if next_chunk is None:
+            break
+        if writer is None:
+            writer = pq.ParquetWriter(out_path, next_chunk.schema)
+        next_table = pa.Table.from_batches([next_chunk])
+        writer.write_table(next_table)
+writer.close()
+```
+
+On my system, the conversion from the CSV to Parquet takes about 190
+seconds. Reading the parquet file can be done with:
+
+```python
+data = pq.ParquetDataset(out_path).read()
+```
+
+With this approach the dataset is in memory, and for this large
+dataset, on my system, I need to be careful with what is running on my
+system (I can't have my web browser open for instance) to be able to
+load this without running out of memory.
+
+## The Dataset API in Python
```

Comment:
This section feels very bare, because it does not mirror the section for R. There's no query, and it leaves us with the view that Pyarrow is just...very bad at this workflow. If it is, then that's unfortunate and hard to talk about meaningfully, but if it does work when you use the dataset API, that's a whole other world, and lets you discuss how the Dataset API is suitable for large files in Python.